pyvista/core/utilities/fileio.py

```diff
@@ -38,7 +38,13 @@
     from pyvista.core.utilities.reader import BaseReader
     from pyvista.plotting.texture import Texture
 
-_VTKWriterAlias = Union[_vtk.vtkXMLWriter, _vtk.vtkDataWriter]
+if TYPE_CHECKING:
+    _VTKWriterAlias = Union[_vtk.vtkXMLWriter, _vtk.vtkDataWriter, _vtk.vtkHDFWriter]
+else:
+    _VTKWriterAlias = Union[_vtk.vtkXMLWriter, _vtk.vtkDataWriter]
+    if hasattr(_vtk, 'vtkHDFWriter'):
+        _VTKWriterAlias = Union[_VTKWriterAlias, _vtk.vtkHDFWriter]
```

Comment:
This else branch is only needed if the types are used at runtime. This is usually needed if using things like `typing.cast`. But, if the type is strictly used for type hints, then we can avoid runtime definitions and avoid the need to handle different vtk versions as a result. Unlike the runtime tests, the type-checking CI only uses the latest VTK version, so if `mypy` is successful on VTK 9.4 then we shouldn't need to bother with supported older versions for type-checking.

Suggest removing branch if runtime behavior is not required
```suggestion
```

Reply:
This requires using `from __future__ import annotations`, which will convert all annotations to strings which can then simply be read and interpreted by type-checkers instead importing the actual objects. All code in PyVista has this import statement.

``` python
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import Sequence

def foo(x: Sequence) -> Sequence: ...

foo([42])
```

Our CI only does static type checking with the latest VTK version. This means that any conditional type-checking branches for older versions will _not_ be checked by CI. This is a maintenance concern and these conditional checks may break in the future without our knowledge. Unless there are compelling reasons to keep these then I think we should remove them.

If you really want to keep this conditional type definition then I suggest moving it fully outside of the `TYPE_CHECKING` block. That way, the runtime unit-tests will provide the necessary coverage (since those test using different VTK versions). You can add a `# noqa` comment in case `ruff` complains about this.

Reply:
To clarify, I think this code as-is would have full test coverage, since there is only one static definition (for vtk9.4) and two runtime definitions (9.4 and <9.4). But, what I'm suggesting is removing the runtime definitions completely. 

Unless we _need_ them at runtime (e.g. if using `typing.cast`), then these definitions can be fully static (i.e. only defined inside the `TYPE_CHECKING` block). And, if we _did_ need them at runtime, then in this case we could remove the static definition (since the type-checking will still pick up all other runtime definitions). So, there is no need to define the type both statically _and_ at runtime.